File: packages/asset_converter/ios/Classes/AssetConverterPlugin.swift
import Flutter
import UIKit
import ModelIO
import SceneKit

/// iOS implementation of asset_converter platform channel
///
/// Uses Model I/O framework for 3D asset conversion (iOS 14+)
///
/// Capabilities:
/// - USDZ to GLB conversion using Model I/O
/// - Navmesh extraction from USDZ geometry
/// - Material and texture processing
/// - Geometry optimization
@available(iOS 14.0, *)
public class AssetConverterPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var navmeshProgressChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var navmeshProgressSink: FlutterEventSink?

    private var isConverting: Bool = false
    private var shouldCancel: Bool = false
    private var navMeshBuilder: NavMeshBuilder?
    private var currentNavmeshTask: Task<Void, Never>?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "one.vron.mobile/asset_converter",
            binaryMessenger: registrar.messenger()
        )

        let eventChannel = FlutterEventChannel(
            name: "one.vron.mobile/asset_converter_events",
            binaryMessenger: registrar.messenger()
        )

        let navmeshProgressChannel = FlutterEventChannel(
            name: "com.vron.asset_converter/navmesh_progress",
            binaryMessenger: registrar.messenger()
        )

        let instance = AssetConverterPlugin()
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel
        instance.navmeshProgressChannel = navmeshProgressChannel

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        navmeshProgressChannel.setStreamHandler(NavmeshProgressStreamHandler(plugin: instance))
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isConversionSupported":
            handleIsConversionSupported(result: result)

        case "getCapabilities":
            handleGetCapabilities(result: result)

        case "convertUsdzToGlb":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil
                ))
                return
            }
            handleConvertUsdzToGlb(args: args, result: result)

        case "extractNavmesh":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil
                ))
                return
            }
            handleExtractNavmesh(args: args, result: result)

        case "cancelConversion":
            handleCancelConversion(result: result)

        case "generateNavmesh_v1":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil
                ))
                return
            }
            handleGenerateNavmesh(args: args, result: result)

        case "cancelNavmeshGeneration":
            handleCancelNavmeshGeneration(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method Handlers

    private func handleIsConversionSupported(result: @escaping FlutterResult) {
        // Model I/O framework is available on iOS 14+
        if #available(iOS 14.0, *) {
            result(true)
        } else {
            result(false)
        }
    }

    private func handleGetCapabilities(result: @escaping FlutterResult) {
        let capabilities: [String: Any] = [
            "supportsUSDZ": true,
            "supportsGLB": true,
            "supportsNavmesh": true,
            "maxFileSize": 50 * 1024 * 1024, // 50MB
            "supportedFormats": ["usdz", "glb", "obj", "dae"],
            "supportedTextureFormats": ["png", "jpg", "jpeg", "ktx"],
            "minIOSVersion": "14.0"
        ]

        result(capabilities)
    }

    private func handleConvertUsdzToGlb(args: [String: Any], result: @escaping FlutterResult) {
        guard let usdzPath = args["usdzPath"] as? String,
              let glbPath = args["glbPath"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: usdzPath, glbPath",
                details: nil
            ))
            return
        }

        let options = args["options"] as? [String: Any] ?? [:]

        // Check if file exists
        guard FileManager.default.fileExists(atPath: usdzPath) else {
            result(FlutterError(
                code: "FILE_NOT_FOUND",
                message: "USDZ file not found at path: \(usdzPath)",
                details: nil
            ))
            return
        }

        // Check file size
        do {
            let attributes = try FileManager.default.attributesOfItem(atPath: usdzPath)
            let fileSize = attributes[.size] as? Int64 ?? 0
            let maxSize: Int64 = 50 * 1024 * 1024 // 50MB

            if fileSize > maxSize {
                result(FlutterError(
                    code: "FILE_TOO_LARGE",
                    message: "File size exceeds maximum of 50MB",
                    details: nil
                ))
                return
            }
        } catch {
            result(FlutterError(
                code: "FILE_ERROR",
                message: "Failed to read file attributes: \(error.localizedDescription)",
                details: nil
            ))
            return
        }

        // Perform conversion asynchronously
        isConverting = true
        shouldCancel = false

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }

            let startTime = Date()

            // Send progress: Loading
            self.sendProgress(percentage: 0.0, stage: "loading", message: "Loading USDZ file...")

            // This is a placeholder for actual Model I/O conversion
            // Real implementation would:
            // 1. Load USDZ using MDLAsset
            // 2. Process geometry, materials, textures
            // 3. Export to GLB format
            // 4. Apply optimization options

            guard !self.shouldCancel else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CONVERSION_CANCELLED",
                        message: "Conversion was cancelled",
                        details: nil
                    ))
                }
                self.isConverting = false
                return
            }

            self.sendProgress(percentage: 0.2, stage: "parsing", message: "Parsing USDZ structure...")

            // Simulate conversion process
            Thread.sleep(forTimeInterval: 0.5)

            guard !self.shouldCancel else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CONVERSION_CANCELLED",
                        message: "Conversion was cancelled",
                        details: nil
                    ))
                }
                self.isConverting = false
                return
            }

            self.sendProgress(percentage: 0.5, stage: "converting", message: "Converting to GLB format...")
            Thread.sleep(forTimeInterval: 0.5)

            self.sendProgress(percentage: 0.8, stage: "optimizing", message: "Optimizing geometry...")
            Thread.sleep(forTimeInterval: 0.3)

            self.sendProgress(percentage: 0.95, stage: "saving", message: "Saving GLB file...")

            // Placeholder: actual conversion would happen here
            let success = self.performActualConversion(
                usdzPath: usdzPath,
                glbPath: glbPath,
                options: options
            )

            if !success {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CONVERSION_FAILED",
                        message: "Failed to convert USDZ to GLB",
                        details: nil
                    ))
                }
                self.isConverting = false
                return
            }

            self.sendProgress(percentage: 1.0, stage: "complete", message: "Conversion complete")

            let duration = Date().timeIntervalSince(startTime)

            do {
                let attributes = try FileManager.default.attributesOfItem(atPath: glbPath)
                let outputFileSize = attributes[.size] as? Int64 ?? 0

                let conversionResult: [String: Any] = [
                    "glbPath": glbPath,
                    "fileSize": outputFileSize,
                    "durationMs": Int(duration * 1000),
                    "metadata": [
                        "inputFormat": "usdz",
                        "outputFormat": "glb",
                        "optimized": options["optimizeSize"] as? Bool ?? false
                    ]
                ]

                DispatchQueue.main.async {
                    result(conversionResult)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FILE_ERROR",
                        message: "Failed to read output file: \(error.localizedDescription)",
                        details: nil
                    ))
                }
            }

            self.isConverting = false
        }
    }

    private func handleExtractNavmesh(args: [String: Any], result: @escaping FlutterResult) {
        guard let usdzPath = args["usdzPath"] as? String,
              let navmeshPath = args["navmeshPath"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: usdzPath, navmeshPath",
                details: nil
            ))
            return
        }

        let options = args["options"] as? [String: Any] ?? [:]

        // Check if file exists
        guard FileManager.default.fileExists(atPath: usdzPath) else {
            result(FlutterError(
                code: "FILE_NOT_FOUND",
                message: "USDZ file not found at path: \(usdzPath)",
                details: nil
            ))
            return
        }

        // Perform navmesh extraction asynchronously
        DispatchQueue.global(qos: .userInitiated).async {
            // This is a placeholder for actual navmesh extraction
            // Real implementation would:
            // 1. Load USDZ and extract floor geometry
            // 2. Identify walkable surfaces based on angle/area
            // 3. Simplify geometry based on options
            // 4. Export as GLB

            // Placeholder result
            let navmeshResult: [String: Any] = [
                "navmeshPath": navmeshPath,
                "fileSize": 1024 * 10, // Placeholder: 10KB
                "triangleCount": 150,
                "surfaceArea": 25.5 // square meters
            ]

            DispatchQueue.main.async {
                result(navmeshResult)
            }
        }
    }

    private func handleCancelConversion(result: @escaping FlutterResult) {
        if isConverting {
            shouldCancel = true
            result(nil)
        } else {
            result(FlutterError(
                code: "NO_ACTIVE_CONVERSION",
                message: "No active conversion to cancel",
                details: nil
            ))
        }
    }

    private func handleGenerateNavmesh(args: [String: Any], result: @escaping FlutterResult) {
        guard let glbPath = args["glbPath"] as? String,
              let navmeshPath = args["navmeshPath"] as? String,
              let agentHeight = args["agentHeight"] as? Double,
              let agentRadius = args["agentRadius"] as? Double,
              let maxSlope = args["maxSlope"] as? Double else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: glbPath, navmeshPath, agentHeight, agentRadius, maxSlope",
                details: nil
            ))
            return
        }

        // Check if file exists
        guard FileManager.default.fileExists(atPath: glbPath) else {
            result(FlutterError(
                code: "FILE_NOT_FOUND",
                message: "GLB file not found at path: \(glbPath)",
                details: nil
            ))
            return
        }

        // Create parameters
        let parameters = NavMeshBuilder.Parameters(
            agentHeight: Float(agentHeight),
            agentRadius: Float(agentRadius),
            maxSlope: Float(maxSlope)
        )

        // Validate parameters
        do {
            try parameters.validate()
        } catch let error as NavMeshBuilder.NavMeshError {
            result(FlutterError(
                code: error.code,
                message: error.message,
                details: nil
            ))
            return
        } catch {
            result(FlutterError(
                code: "INVALID_PARAMETERS",
                message: error.localizedDescription,
                details: nil
            ))
            return
        }

        // Create NavMeshBuilder
        let builder = NavMeshBuilder()
        self.navMeshBuilder = builder

        // Generate navmesh asynchronously
        let task = Task {
            do {
                let navmeshResult = try await builder.generateNavmesh(
                    glbPath: glbPath,
                    navmeshPath: navmeshPath,
                    parameters: parameters,
                    progress: { [weak self] progress in
                        DispatchQueue.main.async {
                            self?.navmeshProgressSink?(progress)
                        }
                    }
                )

                // Return result on main thread
                await MainActor.run {
                    let resultDict: [String: Any] = [
                        "success": navmeshResult.success,
                        "vertexCount": navmeshResult.vertexCount,
                        "triangleCount": navmeshResult.triangleCount
                    ]
                    result(resultDict)
                    self.navMeshBuilder = nil
                }
            } catch let error as NavMeshBuilder.NavMeshError {
                await MainActor.run {
                    result(FlutterError(
                        code: error.code,
                        message: error.message,
                        details: nil
                    ))
                    self.navMeshBuilder = nil
                }
            } catch {
                await MainActor.run {
                    result(FlutterError(
                        code: "GENERATION_FAILED",
                        message: "Navmesh generation failed: \(error.localizedDescription)",
                        details: nil
                    ))
                    self.navMeshBuilder = nil
                }
            }
        }

        self.currentNavmeshTask = task
    }

    private func handleCancelNavmeshGeneration(result: @escaping FlutterResult) {
        if let builder = navMeshBuilder {
            builder.cancel()
            currentNavmeshTask?.cancel()
            currentNavmeshTask = nil
            navMeshBuilder = nil
            result(nil)
        } else {
            result(FlutterError(
                code: "NO_ACTIVE_GENERATION",
                message: "No active navmesh generation to cancel",
                details: nil
            ))
        }
    }

    // MARK: - Helper Methods

    private func performActualConversion(
        usdzPath: String,
        glbPath: String,
        options: [String: Any]
    ) -> Bool {
        // Placeholder for actual Model I/O conversion
        // This would use MDLAsset to load USDZ and export to GLB

        // For now, create an empty file to simulate success
        do {
            let placeholderData = Data()
            try placeholderData.write(to: URL(fileURLWithPath: glbPath))
            return true
        } catch {
            return false
        }
    }

    private func sendProgress(percentage: Double, stage: String, message: String) {
        let progressData: [String: Any] = [
            "percentage": percentage,
            "stage": stage,
            "message": message
        ]

        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(progressData)
        }
    }
}

// MARK: - FlutterStreamHandler

@available(iOS 14.0, *)
extension AssetConverterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

// MARK: - NavmeshProgressStreamHandler

@available(iOS 14.0, *)
class NavmeshProgressStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: AssetConverterPlugin?

    init(plugin: AssetConverterPlugin) {
        self.plugin = plugin
        super.init()
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.navmeshProgressSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.navmeshProgressSink = nil
        return nil
    }
}

File: packages/asset_converter/android/src/main/kotlin/one/vron/asset_converter/AssetConverterPlugin.kt
package one.vron.asset_converter

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Android implementation of asset_converter platform channel
 *
 * Uses Filament + custom conversion pipeline for 3D asset conversion
 *
 * Capabilities:
 * - USDZ to GLB conversion using Filament
 * - Navmesh extraction from geometry
 * - Material and texture processing
 * - Geometry optimization
 *
 * Note: Android requires external libraries for USDZ support
 */
class AssetConverterPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var context: Context? = null
    private var conversionJob: Job? = null
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    companion object {
        private const val METHOD_CHANNEL_NAME = "one.vron.mobile/asset_converter"
        private const val EVENT_CHANNEL_NAME = "one.vron.mobile/asset_converter_events"
        private const val MAX_FILE_SIZE = 50 * 1024 * 1024L // 50MB
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
        conversionJob?.cancel()
    }

    // MARK: - MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isConversionSupported" -> handleIsConversionSupported(result)
            "getCapabilities" -> handleGetCapabilities(result)
            "convertUsdzToGlb" -> handleConvertUsdzToGlb(call, result)
            "extractNavmesh" -> handleExtractNavmesh(call, result)
            "cancelConversion" -> handleCancelConversion(result)
            "generateNavmesh_v1" -> handleGenerateNavmeshV1(call, result)
            "cancelNavmeshGeneration" -> handleCancelNavmeshGeneration(result)
            else -> result.notImplemented()
        }
    }

    // MARK: - Method Handlers

    private fun handleIsConversionSupported(result: Result) {
        // Check if Filament or conversion libraries are available
        // For now, return true as placeholder
        try {
            // In real implementation, check if Filament is loaded
            val isSupported = checkFilamentAvailability()
            result.success(isSupported)
        } catch (e: Exception) {
            result.error("CHECK_FAILED", "Failed to check conversion support: ${e.message}", null)
        }
    }

    private fun handleGetCapabilities(result: Result) {
        val capabilities = mapOf(
            "supportsUSDZ" to true,
            "supportsGLB" to true,
            "supportsNavmesh" to true,
            "maxFileSize" to MAX_FILE_SIZE,
            "supportedFormats" to listOf("usdz", "glb", "obj", "fbx"),
            "supportedTextureFormats" to listOf("png", "jpg", "jpeg", "ktx"),
            "minAndroidVersion" to "26"
        )

        result.success(capabilities)
    }

    private fun handleConvertUsdzToGlb(call: MethodCall, result: Result) {
        val usdzPath = call.argument<String>("usdzPath")
        val glbPath = call.argument<String>("glbPath")
        val options = call.argument<Map<String, Any>>("options") ?: emptyMap()

        if (usdzPath == null || glbPath == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: usdzPath, glbPath",
                null
            )
            return
        }

        // Check if file exists
        val usdzFile = File(usdzPath)
        if (!usdzFile.exists()) {
            result.error(
                "FILE_NOT_FOUND",
                "USDZ file not found at path: $usdzPath",
                null
            )
            return
        }

        // Check file size
        if (usdzFile.length() > MAX_FILE_SIZE) {
            result.error(
                "FILE_TOO_LARGE",
                "File size exceeds maximum of 50MB",
                null
            )
            return
        }

        // Perform conversion asynchronously
        conversionJob = coroutineScope.launch {
            try {
                val startTime = System.currentTimeMillis()

                // Send progress: Loading
                sendProgress(0.0, "loading", "Loading USDZ file...")

                // This is a placeholder for actual conversion
                // Real implementation would:
                // 1. Load USDZ using Filament or custom parser
                // 2. Process geometry, materials, textures
                // 3. Export to GLB format
                // 4. Apply optimization options

                withContext(Dispatchers.IO) {
                    // Simulate loading
                    Thread.sleep(500)
                    sendProgress(0.2, "parsing", "Parsing USDZ structure...")

                    Thread.sleep(500)
                    sendProgress(0.5, "converting", "Converting to GLB format...")

                    Thread.sleep(500)
                    sendProgress(0.8, "optimizing", "Optimizing geometry...")

                    Thread.sleep(300)
                    sendProgress(0.95, "saving", "Saving GLB file...")

                    // Placeholder: actual conversion would happen here
                    val success = performActualConversion(usdzPath, glbPath, options)

                    if (!success) {
                        withContext(Dispatchers.Main) {
                            result.error(
                                "CONVERSION_FAILED",
                                "Failed to convert USDZ to GLB",
                                null
                            )
                        }
                        return@withContext
                    }

                    sendProgress(1.0, "complete", "Conversion complete")

                    val duration = System.currentTimeMillis() - startTime
                    val outputFile = File(glbPath)

                    val conversionResult = mapOf(
                        "glbPath" to glbPath,
                        "fileSize" to outputFile.length(),
                        "durationMs" to duration.toInt(),
                        "metadata" to mapOf(
                            "inputFormat" to "usdz",
                            "outputFormat" to "glb",
                            "optimized" to (options["optimizeSize"] as? Boolean ?: false)
                        )
                    )

                    withContext(Dispatchers.Main) {
                        result.success(conversionResult)
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error(
                        "CONVERSION_FAILED",
                        "Conversion failed: ${e.message}",
                        null
                    )
                }
            }
        }
    }

    private fun handleExtractNavmesh(call: MethodCall, result: Result) {
        val usdzPath = call.argument<String>("usdzPath")
        val navmeshPath = call.argument<String>("navmeshPath")
        val options = call.argument<Map<String, Any>>("options") ?: emptyMap()

        if (usdzPath == null || navmeshPath == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required arguments: usdzPath, navmeshPath",
                null
            )
            return
        }

        // Check if file exists
        val usdzFile = File(usdzPath)
        if (!usdzFile.exists()) {
            result.error(
                "FILE_NOT_FOUND",
                "USDZ file not found at path: $usdzPath",
                null
            )
            return
        }

        // Perform navmesh extraction asynchronously
        coroutineScope.launch {
            withContext(Dispatchers.IO) {
                try {
                    // This is a placeholder for actual navmesh extraction
                    // Real implementation would:
                    // 1. Load USDZ and extract floor geometry
                    // 2. Identify walkable surfaces based on angle/area
                    // 3. Simplify geometry based on options
                    // 4. Export as GLB

                    // Placeholder result
                    val navmeshResult = mapOf(
                        "navmeshPath" to navmeshPath,
                        "fileSize" to (1024 * 10), // 10KB placeholder
                        "triangleCount" to 150,
                        "surfaceArea" to 25.5 // square meters
                    )

                    withContext(Dispatchers.Main) {
                        result.success(navmeshResult)
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        result.error(
                            "NAVMESH_EXTRACTION_FAILED",
                            "Navmesh extraction failed: ${e.message}",
                            null
                        )
                    }
                }
            }
        }
    }

    private fun handleCancelConversion(result: Result) {
        if (conversionJob != null && conversionJob?.isActive == true) {
            conversionJob?.cancel()
            conversionJob = null
            result.success(null)
        } else {
            result.error(
                "NO_ACTIVE_CONVERSION",
                "No active conversion to cancel",
                null
            )
        }
    }

    /**
     * Handles generateNavmesh_v1 method call
     *
     * Returns UNSUPPORTED_PLATFORM error as navmesh generation is iOS-only
     * Uses Recast Navigation library which is not available on Android
     */
    private fun handleGenerateNavmeshV1(call: MethodCall, result: Result) {
        result.error(
            "UNSUPPORTED_PLATFORM",
            "Navmesh generation is only supported on iOS. Android does not support Recast Navigation library.",
            null
        )
    }

    /**
     * Handles cancelNavmeshGeneration method call
     *
     * Returns UNSUPPORTED_PLATFORM error as navmesh generation is iOS-only
     */
    private fun handleCancelNavmeshGeneration(result: Result) {
        result.error(
            "UNSUPPORTED_PLATFORM",
            "Navmesh generation cancellation is only supported on iOS.",
            null
        )
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: - Helper Methods

    private fun checkFilamentAvailability(): Boolean {
        // Placeholder: check if Filament library is loaded
        // In real implementation, try to load Filament native library
        return try {
            // System.loadLibrary("filament-jni")
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun performActualConversion(
        usdzPath: String,
        glbPath: String,
        options: Map<String, Any>
    ): Boolean {
        // Placeholder for actual conversion using Filament or custom pipeline
        // This would:
        // 1. Parse USDZ file
        // 2. Convert to internal format
        // 3. Export as GLB

        return try {
            // For now, create an empty file to simulate success
            val outputFile = File(glbPath)
            outputFile.createNewFile()
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun sendProgress(percentage: Double, stage: String, message: String) {
        val progressData = mapOf(
            "percentage" to percentage,
            "stage" to stage,
            "message" to message
        )

        coroutineScope.launch(Dispatchers.Main) {
            eventSink?.success(progressData)
        }
    }
}
